substrate/genesis-liquidity/pallet/src/lib.rs

```diff
@@ -0,0 +1,471 @@
+          GENESIS_LIQUIDITY_ACCOUNT.into(),
+          Balance { coin: Coin::Serai, amount: Amount(GENESIS_SRI) },
+        )
+        .unwrap();
+
+        // get pool & total values
+        let mut pool_values = vec![];
+        let mut total_value: u128 = 0;
+        for coin in COINS {
+          if coin == Coin::Serai {
+            continue;
+          }
+
+          // initial coin value in terms of btc
+          let Some(value) = Oracle::<T>::get(coin) else {
+            continue;
+          };
+
+          let pool_amount =
+            u128::from(Supply::<T>::get(coin).unwrap_or(LiquidityAmount::zero()).coins);
+          let pool_value = pool_amount
+            .checked_mul(value.into())
+            .unwrap()
+            .checked_div(10u128.pow(coin.decimals()))
+            .unwrap();
+          total_value = total_value.checked_add(pool_value).unwrap();
+          pool_values.push((coin, pool_amount, pool_value));
+        }
+
+        // add the liquidity per pool
+        let mut total_sri_distributed = 0;
+        let pool_values_len = pool_values.len();
+        for (i, (coin, pool_amount, pool_value)) in pool_values.into_iter().enumerate() {
+          // whatever sri left for the last coin should be ~= it's ratio
+          let sri_amount = if i == (pool_values_len - 1) {
+            GENESIS_SRI.checked_sub(total_sri_distributed).unwrap()
+          } else {
+            u64::try_from(
+              u128::from(GENESIS_SRI)
+                .checked_mul(pool_value)
+                .unwrap()
+                .checked_div(total_value)
+                .unwrap(),
+            )
+            .unwrap()
+          };
+          total_sri_distributed = total_sri_distributed.checked_add(sri_amount).unwrap();
+
+          // we can't add 0 liquidity
+          if !(pool_amount > 0 && sri_amount > 0) {
+            continue;
+          }
+
+          // actually add the liquidity to dex
+          let origin = RawOrigin::Signed(GENESIS_LIQUIDITY_ACCOUNT.into());
+          Dex::<T>::add_liquidity(
+            origin.into(),
+            coin,
+            u64::try_from(pool_amount).unwrap(),
+            sri_amount,
+            u64::try_from(pool_amount).unwrap(),
+            sri_amount,
+            GENESIS_LIQUIDITY_ACCOUNT.into(),
+          )
+          .unwrap();
+
+          // let everyone know about the event
+          Self::deposit_event(Event::GenesisLiquidityAddedToPool {
+            coin1: Balance { coin, amount: Amount(u64::try_from(pool_amount).unwrap()) },
+            sri: Amount(sri_amount),
+          });
+        }
+        assert_eq!(total_sri_distributed, GENESIS_SRI);
+
+        // we shouldn't have left any coin in genesis account at this moment, including SRI.
+        // All transferred to the pools.
+        for coin in COINS {
+          assert_eq!(Coins::<T>::balance(GENESIS_LIQUIDITY_ACCOUNT.into(), coin), Amount(0));
+        }
+
+        GenesisComplete::<T>::set(Some(true));
```

Comment:
We can have the value be `()` instead of a `bool`, where if a value was set is if it's complete.